File: scala/ProjectEulerSolutions/src/P0004.scala
package P0004

import Common.Functions.{primeFactors, partitions, fibonacci}
import scala.language.postfixOps
import scala.math.pow

// Scala: Largest palindrome product

// Problem 4
// A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

// Find the largest palindrome made from the product of two 3-digit numbers.
object Solution {
  def largestPalindromeProductOfNDigit(n: BigInt): BigInt = {
    def isPalindrome(m: BigInt): Boolean = m.toString == m.toString.reverse

    def isPalindromeProductOfNDigit(n: BigInt, pali: BigInt): Boolean = {
      val upperBound = BigInt((pow(10, n.toDouble)).toInt)
      val pf         = primeFactors(pali)
      val sets = for {
        p <- partitions(pf)
        if p.size == 2
      } yield p.map(_.product)
      val fs = sets.filter(f => f.head < upperBound && f.tail.head < upperBound)

      fs.map(_.product).filter(_ == pali).distinct.size == 1
    }

    def isPaliAndProductOfNDigit(pali: BigInt): Option[BigInt] =
      if (isPalindrome(pali) && isPalindromeProductOfNDigit(n, pali)) Some(pali)
      else None

    val upperBound    = BigInt((pow(10, n.toDouble) - 1).toInt)
    val lowerBound    = BigInt((pow(10, n.toDouble - 1)).toInt)
    val start: BigInt = upperBound * upperBound
    val end: BigInt   = lowerBound * lowerBound

    LazyList.from[BigInt](start to end by -1) flatMap (isPaliAndProductOfNDigit _ andThen (_.toList)) head
  }
}

File: haskell/project-euler-solutions/src/P0004.hs
-- Largest palindrome product
-- Problem 4
-- A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

-- Find the largest palindrome made from the product of two 3-digit numbers.

module P0004
    ( largestPalindromeProductOfNDigit
    , subsequencesOfSize
    , primeFactors
    , partitions
    , unique
    , isPalindromeProductOfNDigit
    ) where


import Common (maybeHead, primeFactors, partitions, subsequencesOfSize, unique)
import Data.Char(intToDigit)   
import Control.Monad ( filterM )
              

largestPalindromeProductOfNDigit :: Int -> Int
largestPalindromeProductOfNDigit n = head [prod | prod <- [start, start - 1..end], isPalindrome prod && isPalindromeProductOfNDigit n prod]
  where upperBound = 10^n - 1
        lowerBound = 10^(n - 1)
        start = upperBound*upperBound
        end   = lowerBound*lowerBound
        isPalindrome m = show m == reverse (show m)

isPalindromeProductOfNDigit :: Int -> Int -> Bool
isPalindromeProductOfNDigit n pali =  not (null prods) && (maximum prods == pali)
  where upperBound = 10^n
        pf = primeFactors pali
        sets = [map product p | p <- partitions pf, length p == 2]
        fs = filter (\f -> head f < upperBound && head (tail f) < upperBound) sets
        prods = map product fs

